lib/scheduler/constant_arrival_rate.go

```diff
@@ -0,0 +1,73 @@
+package scheduler
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/loadimpact/k6/lib/types"
+	null "gopkg.in/guregu/null.v3"
+)
+
+const constantArrivalRateType = "constant-arrival-rate"
+
+// ConstantArrivalRateConfig stores config for the constant arrival-rate scheduler
+type ConstantArrivalRateConfig struct {
+	BaseConfig
+	Rate     null.Int           `json:"rate"`
+	TimeUnit types.NullDuration `json:"timeUnit"` //TODO: rename to something else?
+	Duration types.NullDuration `json:"duration"`
+
+	// Initialize `PreAllocatedVUs` numeber of VUs, and if more than that are needed,
+	// they will be dynamically allocated, until `MaxVUs` is reached, which is an
+	// absolutely hard limit on the number of VUs the scheduler will use
+	PreAllocatedVUs null.Int `json:"preAllocatedVUs"`
+	MaxVUs          null.Int `json:"maxVUs"`
+}
+
+// NewConstantArrivalRateConfig returns a ConstantArrivalRateConfig with default values
+func NewConstantArrivalRateConfig(name string) ConstantArrivalRateConfig {
+	return ConstantArrivalRateConfig{
+		BaseConfig: NewBaseConfig(name, constantArrivalRateType, false),
+		TimeUnit:   types.NewNullDuration(1*time.Second, false),
+		//TODO: set some default values for PreAllocatedVUs and MaxVUs?
+	}
+}
+
+// Make sure we implement the Config interface
+var _ Config = &ConstantArrivalRateConfig{}
+
+// Validate makes sure all options are configured and valid
+func (carc ConstantArrivalRateConfig) Validate() []error {
```

Comment:
I am pretty sure we will have to either use a validation library or write ourself one. Even if it means adding code generation. Maybe we can leave it until we actually redo the configuration but ... I don't like code like this that will most definitely miss something

Reply:
I looked at several validation libraries, and none of them seemed like a big improvement. The biggest obstacle were the nullable types we use - we'd have to make a bunch of adapters to support them properly, which isn't very worth it... 

I think we can consolidate some of the repeating checks we'll have into helper functions, or reuse some of the validator functions (but not struct tags and reflection-based stuff) from some library for things like IP addresses, urls, etc., but I'm not sure if will make sense to do more than that. Code generation for sure feels an extreme response to this problem... 

And I doubt any other solution will allow us the current flexibility. Consider this validation code below:
```go
if !carc.Rate.Valid {
	errors = append(errors, fmt.Errorf("the iteration rate isn't specified"))
} else if carc.Rate.Int64 <= 0 {
	errors = append(errors, fmt.Errorf("the iteration rate should be positive"))
}
```
Notice how only one of the errors will be specified, since it doesn't make sense to show both the *not specified* and *should be positive* errors at the same time. Not sure how libraries will handle that correctly. Also, we have the flexibility to write very user-friendly and specific error messages. And we can have subtly different validations between the different configs, for example the validation for the variable arrival-rate doesn't require the value and allows it to be 0, but not negative.

I'm sure we can create something with all of the flexibility, but with less boilerplate, it's just likely going to be very difficult, i.e. I'm not sure if it'd end up worth it.